roachtest: restore/tpcc-5k/gce/nodes=4/cpus=8 failed

roachtest.restore/tpcc-5k/gce/nodes=4/cpus=8 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20452357?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20452357?buildTab=artifacts#/restore/tpcc-5k/gce/nodes=4/cpus=8) on master @ [fe3f3462c904d4735fef481fff4536925d9d7a77](https://github.com/cockroachdb/cockroach/commits/fe3f3462c904d4735fef481fff4536925d9d7a77):


```
(monitor.go:267).Wait: monitor failure: context canceled
unexpected node event: n2: cockroach process for system interface died (exit code 134)
test artifacts and logs in: /artifacts/restore/tpcc-5k/gce/nodes=4/cpus=8/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=8</code>
 - <code>encrypted=false</code>
 - <code>fs=ext4</code>
 - <code>localSSD=false</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20452357/restore-tpcc-5k-gce-nodes-4-cpus-8/1757508254321/1757509227243)

</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*restore/tpcc-5k/gce/nodes=4/cpus=8.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54372

We got a panic trace. This is a nested panic. The original panic is bubbling up out of the cloud gcp writer.

```
panic: Release called on a BufferPool with in-use buffers
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait
	  | 	pkg/util/ctxgroup/ctxgroup.go:151
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers
	  | 	pkg/util/ctxgroup/ctxgroup.go:211
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).runRestoreWorkers
	  | 	pkg/backup/restore_data_processor.go:397
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).Start.func3
	  | 	pkg/backup/restore_data_processor.go:229
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).Start.Group.GoCtx.func5
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/pebble/sstable/block.(*BufferPool).Release
	  | 	external/com_github_cockroachdb_pebble/sstable/block/buffer_pool.go:178
	  | github.com/cockroachdb/pebble.(*externalIterState).Close
	  | 	external/com_github_cockroachdb_pebble/external_iterator.go:123
	  | github.com/cockroachdb/pebble.(*Iterator).Close
	  | 	external/com_github_cockroachdb_pebble/iterator.go:2417
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).destroy
	  | 	pkg/storage/pebble_iterator.go:996
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Close
	  | 	pkg/storage/pebble_iterator.go:332
	  | github.com/cockroachdb/cockroach/pkg/storage.(*ReadAsOfIterator).Close
	  | 	pkg/storage/read_as_of_iterator.go:41
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).openSSTs.(*restoreDataProcessor).openSSTs.func2.func3
	  | 	pkg/backup/restore_data_processor.go:334
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
	  | 	pkg/util/tracing/span.go:157
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer
	  | 	pkg/util/tracing/span.go:203
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.ForkSpan
	  | 	pkg/util/tracing/tracer.go:1581
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*loggingTransport).RoundTrip
	  | 	pkg/cloud/cloud_logging_transport.go:54
	  | google.golang.org/api/transport/http.(*parameterTransport).RoundTrip
	  | 	external/org_golang_google_api/transport/http/dial.go:144
	  | go.opencensus.io/plugin/ochttp.(*traceTransport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/trace.go:84
	  | go.opencensus.io/plugin/ochttp.statsTransport.RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client_stats.go:57
	  | go.opencensus.io/plugin/ochttp.(*Transport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client.go:99
	  | golang.org/x/oauth2.(*Transport).RoundTrip
	  | 	external/org_golang_x_oauth2/transport.go:55
	  | net/http.send
	  | 	GOROOT/src/net/http/client.go:259
	  | net/http.(*Client).send
	  | 	GOROOT/src/net/http/client.go:180
	  | net/http.(*Client).do
	  | 	GOROOT/src/net/http/client.go:731
	  | net/http.(*Client).Do
	  | 	GOROOT/src/net/http/client.go:590
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2.1
	  | 	external/com_google_cloud_go_storage/http_client.go:834
	  | cloud.google.com/go/storage.run.func1
	  | 	external/com_google_cloud_go_storage/invoke.go:67
	  | cloud.google.com/go/internal.retry
	  | 	external/com_google_cloud_go/internal/retry.go:40
	  | cloud.google.com/go/internal.Retry
	  | 	external/com_google_cloud_go/internal/retry.go:33
	  | cloud.google.com/go/storage.run
	  | 	external/com_google_cloud_go_storage/invoke.go:65
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2
	  | 	external/com_google_cloud_go_storage/http_client.go:833
	  | cloud.google.com/go/storage.(*httpReader).Read
	  | 	external/com_google_cloud_go_storage/http_client.go:1338
	  | cloud.google.com/go/storage.(*Reader).Read
	  | 	external/com_google_cloud_go_storage/reader.go:207
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*ResumingReader).Read
	  | 	pkg/cloud/cloud_io.go:319
	Wraps: (3) assertion failure
	Wraps: (4) attached stack trace
	  -- stack trace:
	  | github.com/cockroachdb/pebble/sstable/block.(*BufferPool).Release
	  | 	external/com_github_cockroachdb_pebble/sstable/block/buffer_pool.go:178
	  | github.com/cockroachdb/pebble.(*externalIterState).Close
	  | 	external/com_github_cockroachdb_pebble/external_iterator.go:123
	  | github.com/cockroachdb/pebble.(*Iterator).Close
	  | 	external/com_github_cockroachdb_pebble/iterator.go:2417
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).destroy
	  | 	pkg/storage/pebble_iterator.go:996
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Close
	  | 	pkg/storage/pebble_iterator.go:332
	  | github.com/cockroachdb/cockroach/pkg/storage.(*ReadAsOfIterator).Close
	  | 	pkg/storage/read_as_of_iterator.go:41
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).openSSTs.(*restoreDataProcessor).openSSTs.func2.func3
	  | 	pkg/backup/restore_data_processor.go:334
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
	  | 	pkg/util/tracing/span.go:157
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer
	  | 	pkg/util/tracing/span.go:203
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.ForkSpan
	  | 	pkg/util/tracing/tracer.go:1581
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*loggingTransport).RoundTrip
	  | 	pkg/cloud/cloud_logging_transport.go:54
	  | google.golang.org/api/transport/http.(*parameterTransport).RoundTrip
	  | 	external/org_golang_google_api/transport/http/dial.go:144
	  | go.opencensus.io/plugin/ochttp.(*traceTransport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/trace.go:84
	  | go.opencensus.io/plugin/ochttp.statsTransport.RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client_stats.go:57
	  | go.opencensus.io/plugin/ochttp.(*Transport).RoundTrip
	   | 	external/io_opencensus_go/plugin/ochttp/client.go:99
	  | golang.org/x/oauth2.(*Transport).RoundTrip
	  | 	external/org_golang_x_oauth2/transport.go:55
	  | net/http.send
	  | 	GOROOT/src/net/http/client.go:259
	  | net/http.(*Client).send
	  | 	GOROOT/src/net/http/client.go:180
	  | net/http.(*Client).do
	  | 	GOROOT/src/net/http/client.go:731
	  | net/http.(*Client).Do
	  | 	GOROOT/src/net/http/client.go:590
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2.1
	  | 	external/com_google_cloud_go_storage/http_client.go:834
	  | cloud.google.com/go/storage.run.func1
	  | 	external/com_google_cloud_go_storage/invoke.go:67
	  | cloud.google.com/go/internal.retry
	  | 	external/com_google_cloud_go/internal/retry.go:40
	  | cloud.google.com/go/internal.Retry
	  | 	external/com_google_cloud_go/internal/retry.go:33
	  | cloud.google.com/go/storage.run
	  | 	external/com_google_cloud_go_storage/invoke.go:65
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2
	  | 	external/com_google_cloud_go_storage/http_client.go:833
	  | cloud.google.com/go/storage.(*httpReader).Read
	  | 	external/com_google_cloud_go_storage/http_client.go:1338
	  | cloud.google.com/go/storage.(*Reader).Read
	  | 	external/com_google_cloud_go_storage/reader.go:207
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*ResumingReader).Read
	  | 	pkg/cloud/cloud_io.go:319
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*limitedReader).Read
	  | 	pkg/cloud/impl_registry.go:443
	  | github.com/cockroachdb/cockroach/pkg/ccl/storageccl.(*sstReader).ReadAt
	  | 	pkg/ccl/storageccl/external_sst_reader.go:287
	Wraps: (5) Release called on a BufferPool with in-use buffers
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) *assert.withAssertionFailure (4) *withstack.withStack (5) *errutil.leafError
```

--- a/pkg/cloud/azure/azure_storage.go
+++ b/pkg/cloud/azure/azure_storage.go
@@ -366,6 +366,7 @@ func (s *azureStorage) ReadFile(
 			}
 		}
 	}
+	// BUG: we should follow the azure retry setting here.
 	reader := resp.NewRetryReader(ctx, &azblob.RetryReaderOptions{MaxRetries: 3})
 	return ioctx.ReadCloserAdapter(reader), fileSize, nil
 }

--- a/pkg/cloud/cloud_io.go
+++ b/pkg/cloud/cloud_io.go
@@ -242,6 +242,7 @@ type ReaderOpenerAt func(ctx context.Context, pos int64) (io.ReadCloser, int64, 
 type ResumingReader struct {
 	Opener       ReaderOpenerAt   // Get additional content
 	Reader       io.ReadCloser    // Currently opened reader
+	ReaderSpan   *tracing.Span    // Span for the current reader, if Reader is non-nil
 	Filename     string           // Used for logging
 	Pos          int64            // How much data was received so far
 	Size         int64            // Total size of the file
@@ -284,6 +285,10 @@ func NewResumingReader(
 
 // Open opens the reader at its current offset.
 func (r *ResumingReader) Open(ctx context.Context) error {
+	if r.Reader != nil {
+		return errors.AssertionFailedf("reader already open")
+	}
+
 	if r.Size > 0 && r.Pos >= r.Size {
 		// Don't try to open a file if the size has been set and the position is
 		// at size. This generally results in an invalid range error for the
@@ -294,10 +299,18 @@ func (r *ResumingReader) Open(ctx context.Context) error {
 
 	return DelayedRetry(ctx, "Open", r.ErrFn, func() error {
 		var readErr error
+
+		ctx, span := tracing.ForkSpan(ctx, "resuming-reader")
 		r.Reader, r.Size, readErr = r.Opener(ctx, r.Pos)
 		if readErr != nil {
+			span.Finish()
 			return errors.Wrapf(readErr, "open %s", r.Filename)
 		}
+
+		// We hold onto the span for the lifetime of the reader because the reader
+		// may issue new HTTP requests after Open returns.
+		r.ReaderSpan = span
+
 		return nil
 	})
 }
@@ -340,10 +353,9 @@ func (r *ResumingReader) Read(ctx context.Context, p []byte) (int, error) {
 			}
 			log.Dev.Errorf(ctx, "Retry IO error: %s", lastErr)
 			lastErr = nil
-			if r.Reader != nil {
-				r.Reader.Close()
-			}
-			r.Reader = nil
+			// Ignore the error from Close(). We are already handling a read error
+			// so we know the handle is in a bad state.
+			_ = r.Close(ctx)
 		}
 	}
 
@@ -356,10 +368,14 @@ func (r *ResumingReader) Read(ctx context.Context, p []byte) (int, error) {
 
 // Close implements io.Closer.
 func (r *ResumingReader) Close(ctx context.Context) error {
-	if r.Reader != nil {
-		return r.Reader.Close()
+	if r.Reader == nil {
+		return nil
 	}
-	return nil
+
+	err := r.Reader.Close()
+	r.ReaderSpan.Finish()
+	r.Reader = nil
+	return err
 }
 
 // CheckHTTPContentRangeHeader parses Content-Range header and ensures that

--- a/pkg/cloud/cloudtestutils/BUILD.bazel
+++ b/pkg/cloud/cloudtestutils/BUILD.bazel
@@ -23,6 +23,8 @@ go_library(
         "//pkg/util/syncutil",
         "//pkg/util/sysutil",
         "//pkg/util/timeutil",
+        "//pkg/util/tracing",
+        "//pkg/util/tracing/tracingpb",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_stretchr_testify//require",
         "@org_golang_x_sync//errgroup",

--- a/pkg/cloud/cloudtestutils/cloud_nemesis.go
+++ b/pkg/cloud/cloudtestutils/cloud_nemesis.go
@@ -19,6 +19,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/cloud"
 	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/sync/errgroup"
@@ -39,9 +41,15 @@ func RunCloudNemesisTest(t *testing.T, storage cloud.ExternalStorage) {
 		listConcurrency:  1,
 	}
 
-	// We create a context here because we don't want to support a caller supplied
-	// cancelation signal.
-	ctx := context.Background()
+	// Create a root tracing span for the test to ensure ops create spans
+	// which will help flush out span lifetime bugs.
+	tracer := tracing.NewTracerWithOpt(
+		context.Background(),
+		tracing.WithUseAfterFinishOpt(true, true),
+		tracing.WithSpanReusePercent(0))
+	ctx, rootSpan := tracer.StartSpanCtx(context.Background(), "cloud-nemesis-test", tracing.WithRecording(tracingpb.RecordingStructured))
+	defer rootSpan.Finish()
+
 	if err := nemesis.run(ctx, 2*time.Minute); err != nil {
 		t.Fatalf("%+v", err)
 	}
@@ -85,9 +93,13 @@ func (c *cloudNemesis) run(ctx context.Context, duration time.Duration) error {
 	g, ctx := errgroup.WithContext(ctx)
 
 	g.Go(func() error {
-		time.Sleep(duration)
+		wait := time.NewTimer(duration)
+		select {
+		case <-ctx.Done():
+		case <-wait.C:
+		}
 		close(done)
-		return nil
+		return ctx.Err()
 	})
 
 	for i := 0; i < c.writeConcurrency; i++ {

